File: dynamic_programming/projects/projects.py
from sys import stdin

# Works but just barely TLE on CSES, use C++

def max_projects(projects):
    projects.sort(key=lambda x: x[1])
    dp = [0 for _ in range(len(projects))]
    for i in range(len(dp)):
        start, _, money = projects[i]
        index = find_latest(projects, start, i - 1) if i else -1
        dp[i] = max(dp[i - 1], money if index == -1 else money + dp[index])
    return dp[-1]

def find_latest(projects, start_time, e):
    s = 0
    while s < e:
        mid = -(-(s + e) // 2)
        if projects[mid][1] >= start_time:
            e = mid - 1
        else:
            s = mid
    return s if projects[s][1] < start_time else -1        

def read_num():
    return int(stdin.readline())


if __name__ == '__main__':
    _ = read_num()
    projects = []
    for line in stdin.readlines():
        projects.append(tuple([int(num) for num in line.split()]))
    print(max_projects(projects))

File: dynamic_programming/projects/projects.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>
using namespace std;

// TODO get this working

/*

int comp_projects(tuple<int, int, int> p1, tuple<int, int, int> p2) {
    return get<1>(p1) < get<2>(p2);
}

int find_index(vector<tuple<int, int, int> > projects, int start_time, int e) {
    int s = 0;
    while (s < e) {
        int mid = (s + e) / 2 + ((s + e) % 2 != 0);
        if (get<1>(projects[mid]) >= start_time) {
            e = mid - 1; 
        } else {
            s = mid;
        }
    }
    return get<1>(projects[s]) < start_time ? s : -1; 
}


long long max_projects(vector<tuple<int, int, int> > projects) {
    vector<long long> dp(projects.size());
    sort(projects.begin(), projects.end(), comp_projects);
    for (int i = 1; i < projects.size(); i++) {
        int start, end, money;
        tie(start, end, money) = projects[i];
        int last_project_index = find_index(projects, start, i - 1);
        dp[i] = max(dp[i - 1], last_project_index == -1 ? money : money + dp[last_project_index]);
    }
    return dp[dp.size() - 1];
}

int read_num() {
    int num;
    scanf("%d", &num);
    return num;
}

int main() {
    int n = read_num();
    vector<tuple<int, int, int> > projects;
    while (n--) {
        projects.push_back(make_tuple(read_num(), read_num(), read_num()));
    }
    cout << max_projects(projects) << endl;
}

*/